File: Tree/#226. Invert Binary Tree.cpp

//Using Stack
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        stack<TreeNode*>st;
        st.push(root);
        
        while(!st.empty()){
            auto cur = st.top(); st.pop();
            if(cur){
                st.push(cur->left);
                st.push(cur->right);
                swap(cur->left,cur->right);
            }
        }
        return root;
    }
};


//recursion
class Solution {
public:
    
    TreeNode* helper(TreeNode* root){
       if(!root) return NULL;
        TreeNode* left = root->left;
        root->left = helper(root->right);
        root->right = helper(left);
        return root;
    }
    
    TreeNode* invertTree(TreeNode* root) {
        return helper(root);
    }
};


//Another solution
void invert(Tree* root){
    if(!root) return;
    invert(root->left);
    invert(root->right);
    Tree* leftChild = root->left;
    root->left = root->right;
    root->right = leftChild;
}
Tree* solve(Tree* root) {
    invert(root);
    return root;
}

File: Tree/#226. Invert Binary Tree.py
'''
TC: O(n)
SC: O(n) for call stacks
'''
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None

        root.left = self.invertTree(root.left)
        root.right = self.invertTree(root.right)
        root.left, root.right = root.right, root.left
        return root